src/Shared/FileMatcher.cs

```diff
@@ -705,128 +758,181 @@ private static void GetFilesRecursive
+                            newSearchesToExclude.Add(thisExcludeStep);
                         }
                     }
+                }
 
-                    if (searchesToExcludeInSubdirs != null)
-                    {
-                        List<RecursionState> searchesForSubdir;
+                if (searchesToExcludeInSubdirs != null)
+                {
+                    List<RecursionState> searchesForSubdir;
 
-                        if (searchesToExcludeInSubdirs.TryGetValue(subdir, out searchesForSubdir))
+                    if (searchesToExcludeInSubdirs.TryGetValue(subdir, out searchesForSubdir))
+                    {
+                        //  We've found the base directory that these exclusions apply to.  So now add them as normal searches
+                        if (newSearchesToExclude == null)
                         {
-                            //  We've found the base directory that these exclusions apply to.  So now add them as normal searches
-                            if (newSearchesToExclude == null)
-                            {
-                                newSearchesToExclude = new List<RecursionState>();
-                            }
-                            newSearchesToExclude.AddRange(searchesForSubdir);
+                            newSearchesToExclude = new List<RecursionState>();
                         }
+                        newSearchesToExclude.AddRange(searchesForSubdir);
                     }
+                }
+
+                // We never want to strip the project directory from the leaves, because the current 
+                // process directory maybe different
+                GetFilesRecursive(
+                    listOfFiles,
+                    newRecursionState,
+                    projectDirectory,
+                    stripProjectDirectory,
+                    getFileSystemEntries,
+                    newSearchesToExclude,
+                    searchesToExcludeInSubdirs,
+                    taskOptions);
+            };
 
-                    // We never want to strip the project directory from the leaves, because the current 
-                    // process directory maybe different
-                    GetFilesRecursive(
-                        listOfFiles,
-                        newRecursionState,
-                        projectDirectory,
-                        stripProjectDirectory,
-                        getFileSystemEntries,
-                        newSearchesToExclude,
-                        searchesToExcludeInSubdirs);
+            // Calcuate the MaxDegreeOfParallelism value in order to prevent too much tasks being running concurrently.
+            int dop = 0;
+            // Lock only when we may be dealing with multiple threads
+            if (taskOptions.MaxTasks > 1 && taskOptions.MaxTasksPerIteration > 1)
+            {
+                lock (s_directoryLock)
```

Comment:
Can you explain what you are protecting against?

Reply:
.. and rename the variable to reflect that. (s_taskOptionsLock?)

Reply:
This lock prevents multiple threads that were created from different `Parallel.ForEach` calls from modifying the `dop` variable in order to assure, that we are not going to use more tasks that have been specified in `TaskOptions`. We have to consider that this method is recursive, which means that we could have one `Parallel.ForEach` that may call another `Parallel.ForEach` inside its iteration. [I've added](https://github.com/Microsoft/msbuild/pull/2572/commits/0fec532e6551c12008ed12b0dc223736edf354a0) an additional condition for locking, as with the current settings the lock is not needed. To demonstrate how this works, I've created [a simulation](https://dotnetfiddle.net/GVvg02), where `TaskOptions` is configured to have up to 4 `Parallel.ForEach` running concurrently. By commenting the `lock` statement, we can see that after several runs, we may get 5 `Parallel.ForEach` running concurrently.

Reply:
👍

Reply:
Can we just lock the taskoptions instead of creating another obhect?

Reply:
Good idea! By doing so, we will have the locking per call, which is more correct than having it static, as the purpose of the locking is just to modify the `TaskOptions`.